File: Data Structures & Algorithms/subtree-of-a-binary-tree/submission-5.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:   
    def isSubtree(self, root: Optional[TreeNode], subRoot: Optional[TreeNode]) -> bool:
        if not root and not subRoot:
            return True
        if not root or not subRoot:
            return False
        
        def dfs(node1, node2):
            if not node1 and not node2:
                return True
            if not node1 or not node2:
                return False
            if node1.val != node2.val:
                return False

            return dfs(node1.left, node2.left) and \
                    dfs(node1.right, node2.right)

        que = deque([root])

        while que:
            for _ in range(len(que)):
                node = que.popleft()
                if node.val == subRoot.val:
                    if dfs(node, subRoot):
                        return True
                if node.left:
                    que.append(node.left)
                if node.right:
                    que.append(node.right)
            
        return False

File: Data Structures & Algorithms/lowest-common-ancestor-in-binary-search-tree/submission-5.js
/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} root
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {TreeNode}
     */
    lowestCommonAncestor(root, p, q) {

        function walk(node, target) {
            if (!node) return [];
            if (node.val == target) return [node];

            const left = walk(node.left, target);
            const right = walk(node.right, target);

            if (left.length > 0) {
                left.push(node);
                return left
            } else if (right.length > 0) {
                right.push(node);
                return right;
            }  else {
                return [];
            }
        }
        // they say p and q will exist in the tree so I am not checking whether they are null or not
        const pPath = walk(root, p.val).reverse();
        const qPath = walk(root, q.val).reverse();

        let i = 0;
        while (i < Math.min(pPath.length, qPath.length) && pPath[i].val == qPath[i].val) i++;
        return pPath[i - 1];
        // if (i - 1 < pPath.length) return  pPath[i - 1];
        // else return qPath[i - 1];
    }
}
